README.md
# Json Table Widget [![GitHub stars](https://img.shields.io/github/stars/apgapg/json_table.svg?style=social)](https://github.com/apgapg/json_table) [![Twitter Follow](https://img.shields.io/twitter/url/https/@ayushpgupta.svg?style=social)](https://twitter.com/ayushpgupta) ![GitHub last commit](https://img.shields.io/github/last-commit/apgapg/json_table.svg) [![Website shields.io](https://img.shields.io/website-up-down-green-red/http/shields.io.svg)](https://play.google.com/store/apps/details?id=com.coddu.flutterprofile)[![Open Source Love](https://badges.frapsoft.com/os/v2/open-source.svg?v=103)](https://github.com/apgapg/json_table)


This Flutter package provides a Json Table Widget for directly showing table from a json(Map). Supports Column toggle also.

Live Demo: [https://apgapg.github.io/json_table/](https://apgapg.github.io/json_table/)

Live Data Testing: [https://apgapg.github.io/json_table/#/customData](https://apgapg.github.io/json_table/#/customData)

### Features
- The table constructed isn't the flutter's native DataTable.
- The table is manually coded hence serves a great learning purpose on how to create simple tables manually in flutter
- Supports vertical & horizontal scroll
- Supports custom columns includes default value, column name, value builder
- Supports nested data showing
- Supports pagination
- Supports row select color, callback
  
  

<img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/s1.gif"  height = "400" alt="JsonTable"> <img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/s2.gif"  height = "400" alt="JsonTable"> <img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/s3.gif"  height = "400" alt="JsonTable"> <img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/ss1.png"  height = "400" alt="JsonTable"> <img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/ss4.png"  height = "400" alt="JsonTable"> <img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/ss6.png"  height = "400" alt="JsonTable">


# 💻 Installation
In the `dependencies:` section of your `pubspec.yaml`, add the following line:

[![Version](https://img.shields.io/pub/v/json_table.svg)](https://pub.dartlang.org/packages/json_table)

```yaml
dependencies:
  json_table: <latest version>
```

# ❔ Usage

### Import this class

```dart
import 'package:json_table/json_table.dart';
```

### - Vanilla Implementation
```dart
//Decode your json string
final String jsonSample='[{"id":1},{"id":2}]';
var json = jsonDecode(jsonSample);

//Simply pass this json to JsonTable
child: JsonTable(json)
```

### - Implementation with HEADER and CELL widget builders
```dart
JsonTable(
   json,
   tableHeaderBuilder: (String header) {
     return Container(
       padding: EdgeInsets.symmetric(horizontal: 8.0, vertical: 4.0),
       decoration: BoxDecoration(border: Border.all(width: 0.5),color: Colors.grey[300]),
       child: Text(
         header,
         textAlign: TextAlign.center,
         style: Theme.of(context).textTheme.display1.copyWith(fontWeight: FontWeight.w700, fontSize: 14.0,color: Colors.black87),
       ),
     );
   },
   tableCellBuilder: (value) {
     return Container(
       padding: EdgeInsets.symmetric(horizontal: 4.0, vertical: 2.0),
       decoration: BoxDecoration(border: Border.all(width: 0.5, color: Colors.grey.withOpacity(0.5))),
       child: Text(
         value,
         textAlign: TextAlign.center,
         style: Theme.of(context).textTheme.display1.copyWith(fontSize: 14.0, color: Colors.grey[900]),
       ),
     );
   },
 )
```

Head over to example code: [simple_table.dart](https://github.com/apgapg/json_table/blob/master/example/lib/pages/simple_table.dart)

### - Implementation with custom COLUMNS list   

- Pass custom column list to control what columns are displayed in table
- The list item must be constructed using JsonTableColumn class
- JsonTableColumn provides 4 parameters, namely,
```dart
JsonTableColumn("age", label: "Eligible to Vote", valueBuilder: eligibleToVote, defaultValue:"NA")
```
- First parameter is the field/key to pick from the data
- label: The column header label to be displayed
- defaultValue: To be used when data or key is null
- valueBuilder: To get the formatted value like date etc

<img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/ss1.png"  height = "400" alt="JsonTable"> 
      
```dart
//Decode your json string
final String jsonSample='[{"name":"Ram","email":"ram@gmail.com","age":23,"DOB":"1990-12-01"},'
                              '{"name":"Shyam","email":"shyam23@gmail.com","age":18,"DOB":"1995-07-01"},'
                              '{"name":"John","email":"john@gmail.com","age":10,"DOB":"2000-02-24"},'
                              '{"name":"Ram","age":12,"DOB":"2000-02-01"}]';
var json = jsonDecode(jsonSample);
//Create your column list
var columns = [
      JsonTableColumn("name", label: "Name"),
      JsonTableColumn("age", label: "Age"),
      JsonTableColumn("DOB", label: "Date of Birth", valueBuilder: formatDOB),
      JsonTableColumn("age", label: "Eligible to Vote", valueBuilder: eligibleToVote),
      JsonTableColumn("email", label: "E-mail", defaultValue: "NA"),
    ];
//Simply pass this column list to JsonTable
child: JsonTable(json,columns: columns)

//Example of valueBuilder
String eligibleToVote(value) {
    if (value >= 18) {
      return "Yes";
    } else
      return "No";
}
```

Head over to example code: [custom_column_table.dart](https://github.com/apgapg/json_table/blob/master/example/lib/pages/custom_column_table.dart)

### - Implementation with nested data list   

Suppose your json object has nested data like email as shown below:
```dart
{"name":"Ram","email":{"1":"ram@gmail.com"},"age":23,"DOB":"1990-12-01"}
```
- Just use email.1 instead of email as key
```dart
JsonTableColumn("email.1", label: "Email", defaultValue:"NA")
```

<img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/ss3.png"  height = "400" alt="JsonTable"> 
      
```dart
//Decode your json string
final String jsonSample='[{"name":"Ram","email":{"1":"ram@gmail.com"},"age":23,"DOB":"1990-12-01"},'
                               '{"name":"Shyam","email":{"1":"shyam23@gmail.com"},"age":18,"DOB":"1995-07-01"},'
                               '{"name":"John","email":{"1":"john@gmail.com"},"age":10,"DOB":"2000-02-24"}]';
var json = jsonDecode(jsonSample);
//Create your column list
var columns = [
      JsonTableColumn("name", label: "Name"),
      JsonTableColumn("age", label: "Age"),
      JsonTableColumn("DOB", label: "Date of Birth", valueBuilder: formatDOB),
      JsonTableColumn("age", label: "Eligible to Vote", valueBuilder: eligibleToVote),
      JsonTableColumn("email.1", label: "E-mail", defaultValue: "NA"),
    ];
//Simply pass this column list to JsonTable
child: JsonTable(json,columns: columns)
```

Head over to example code: [custom_column_nested_table.dart](https://github.com/apgapg/json_table/blob/master/example/lib/pages/custom_column_nested_table.dart)

### Column toggle

Option for toggling column(s) also. User can customise which columns are to be shown

```dart
 showColumnToggle: true
```

<img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/ss5.png"  height = "400" alt="JsonTable"> 

### Row Highlighting

Add row highlighting with custom color support

<img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/ss4.png"  height = "400" alt="JsonTable"> 

```dart
allowRowHighlight: true,
rowHighlightColor: Colors.yellow[500].withOpacity(0.7),
```

### Row Select Callback

Get the index and data map of a particular selected row. 
Note index might return incorrect value in case of pagination

```dart
onRowSelect: (index, map) {
    print(index);
    print(map);
  },
```

### Pagination

Just provide an int value to `paginationRowCount` parameter

<img src="https://raw.githubusercontent.com/apgapg/json_table/master/src/ss6.png"  height = "400" alt="JsonTable"> 

```dart
paginationRowCount: 4,
```

## TODO
- [X] Custom header list parameter. This will help to show only those keys as mentioned in header list
- [X] Add support for keys missing in json object
- [X] Add support for auto formatting of date
- [X] Extracting column headers logic must be change. Not to depend on first object
- [X] Nested data showing support
- [X] Row highlight support
- [X] Row select callback
- [X] Wrap filters in expansion tile
- [X] Pagination support
- [ ] Add option to change header row to vertical row on left

# ⭐ My Flutter Packages
- [pie_chart](https://pub.dartlang.org/packages/pie_chart)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/pie_chart.svg?style=social)](https://github.com/apgapg/pie_chart)  Flutter Pie Chart with cool animation.
- [avatar_glow](https://pub.dartlang.org/packages/avatar_glow)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/avatar_glow.svg?style=social)](https://github.com/apgapg/avatar_glow)  Flutter Avatar Glow Widget with glowing animation.
- [search_widget](https://pub.dartlang.org/packages/search_widget)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/search_widget.svg?style=social)](https://github.com/apgapg/search_widget)  Flutter Search Widget for selecting an option from list.
- [animating_location_pin](https://pub.dev/packages/animating_location_pin)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/animating_location_pin.svg?style=social)](https://github.com/apgapg/animating_location_pin)  Flutter Animating Location Pin Widget providing Animating Location Pin Widget which can be used while fetching device location.
                                                                                                                                                                                                                             
# ⭐ My Flutter Apps
- [flutter_profile](https://github.com/apgapg/flutter_profile)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/flutter_profile.svg?style=social)](https://github.com/apgapg/flutter_profile)  Showcase My Portfolio: Ayush P Gupta on Playstore.
- [flutter_sankalan](https://github.com/apgapg/flutter_sankalan)  [![GitHub stars](https://img.shields.io/github/stars/apgapg/flutter_sankalan.svg?style=social)](https://github.com/apgapg/flutter_sankalan)  Flutter App which allows reading/uploading short stories.

# 👍 Contribution
1. Fork it
2. Create your feature branch (git checkout -b my-new-feature)
3. Commit your changes (git commit -m 'Add some feature')
4. Push to the branch (git push origin my-new-feature)
5. Create new Pull Request


CHANGELOG.md
## [2.0.1]
- Remove json_utilities dependency (#36)

## [2.0.0]
- [Null Safety] Migrate project to support null safety
- [Misc] Update Android and Linux project

## [1.4.0]
- Fixed pagination leaving out elements
- Fixed duplicate item issue
- Restructure project
- Add onRowSelect support
- Add local json showing example

## [1.3.2]
- Fix Pagination not working

## [1.3.1]
- Fix NPE when paginationRowCount is not supplied

## [1.3.0]
- Add pagination support [#1](https://github.com/apgapg/json_table/issues/1)
- Fix column toggle not working when custom column list is provided

## [1.2.0]
- Wrap column filters in ExpansionTile
- Add rowHighlight support

## [1.1.0]
- Add nested data showing support

## [1.0.1]
- Add json decode support for json column
- Fix header text color to black irrespective of theme

## [1.0.0]
- Remove column toggle when column list is provided
- Fix text colors for dark theme

## [0.4.1]
- Reformat code with flutter format

## [0.4.0] - Support Column toggle also.
- A bool as a parameter is provided to toggle column visibility

## [0.3.0] - Change keys extraction code  
- Keys extraction is no more depended on first object of array. All possible keys in source data is taken into account

## [0.2.0] - Custom Columns Added  
- Add Custom Columns support for controlling what data is to be shown

## [0.1.0] - Initial Release


lib/src/table_column.dart
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

import '../json_table.dart';
import 'json_table.dart';
import 'json_table_column.dart';
import 'json_utilities.dart';

class TableColumn extends StatelessWidget {
  final String? header;
  final List? dataList;
  final TableHeaderBuilder? tableHeaderBuilder;
  final TableCellBuilder? tableCellBuilder;
  final JsonTableColumn? column;
  final jsonUtils = JSONUtils();
  final Function(int index, dynamic rowMap) onRowTap;
  final int? highlightedRowIndex;
  final bool allowRowHighlight;
  final Color? rowHighlightColor;

  TableColumn(
    this.header,
    this.dataList,
    this.tableHeaderBuilder,
    this.tableCellBuilder,
    this.column,
    this.onRowTap,
    this.highlightedRowIndex,
    this.allowRowHighlight,
    this.rowHighlightColor,
  );

  @override
  Widget build(BuildContext context) {
    return IntrinsicWidth(
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: <Widget>[
          tableHeaderBuilder != null
              ? tableHeaderBuilder!(header)
              : Container(
                  padding: EdgeInsets.symmetric(horizontal: 4.0, vertical: 2.0),
                  decoration: BoxDecoration(
                    border: Border.all(width: 0.5),
                    color: Colors.grey[300],
                  ),
                  child: Text(
                    header!,
                    textAlign: TextAlign.center,
                    style: TextStyle(
                      fontWeight: FontWeight.w700,
                      fontSize: 14.0,
                      color: Colors.black87,
                    ),
                  ),
                ),
          Container(
            child: Column(
              children: dataList!
                  .map(
                    (rowMap) => Column(
                      crossAxisAlignment: CrossAxisAlignment.stretch,
                      children: <Widget>[
                        GestureDetector(
                          onTap: () {
                            onRowTap(dataList!.indexOf(rowMap), rowMap);
                          },
                          child: Container(
                            color: (allowRowHighlight &&
                                    highlightedRowIndex != null &&
                                    highlightedRowIndex ==
                                        dataList!.indexOf(rowMap))
                                ? rowHighlightColor ??
                                    Colors.yellowAccent.withOpacity(0.7)
                                : null,
                            child: tableCellBuilder != null
                                ? tableCellBuilder!(
                                    getFormattedValue(
                                      jsonUtils.get(
                                        rowMap,
                                        column?.field ?? header!,
                                        column?.defaultValue ?? '',
                                      ),
                                      column,
                                    ),
                                  )
                                : Container(
                                    padding: EdgeInsets.symmetric(
                                        horizontal: 4.0, vertical: 2.0),
                                    decoration: BoxDecoration(
                                        border: Border.all(
                                      width: 0.5,
                                      color: Colors.grey.withOpacity(0.5),
                                    )),
                                    child: Text(
                                      getFormattedValue(
                                        jsonUtils.get(
                                          rowMap,
                                          column?.field ?? header!,
                                          column?.defaultValue ?? '',
                                        ),
                                        column,
                                      ),
                                      textAlign: TextAlign.center,
                                      style: TextStyle(
                                        fontSize: 14.0,
                                      ),
                                    ),
                                  ),
                          ),
                        ),
                      ],
                    ),
                  )
                  .toList(),
            ),
          )
        ],
      ),
    );
  }

  String getFormattedValue(dynamic value, JsonTableColumn? column) {
    if (column != null && column.type != null && column.type!.isNotEmpty) {
      try {
        if (column.type == 'date') {
          return DateFormat("dd MMM yy").format(DateTime.parse(value));
        } else if (column.type == 'time') {
          return DateFormat("hh:mm a").format(DateTime.parse(value));
        } else if (column.type == 'dateTime') {
          return DateFormat("dd MMM yy hh:mm a").format(DateTime.parse(value));
        } else {
          return value.toString();
        }
      } catch (e) {
        return value.toString();
      }
    }
    if (value == null) return column?.defaultValue ?? '';
    if (column?.valueBuilder != null) {
      return column!.valueBuilder!(value);
    }
    return value.toString();
  }
}
